### PR TITLE
Adds seqable operator

### PR DESCRIPTION
**Pull request by jimmyhmiller**
`seqable` initially grew out of my goal to git rid of the comments on scan and vscan about a useless `seq?` check. scan and vscan coerced things using seq, so really they operate on any seqable.

This is where things started and technically it could stop there. I could have made this a purely internal change. But as I thought more about it, I thought this might be a useful operator.

First, it is has already been implicitly asked for. In slack the question of why meander matches on collection type was asked. I personally agree with the justification given, I think this is the right default for meander. At the same time, meander is made to fit into existing code and often that code will produce various seqable collections and it might be nice to express this fact.

`seqable` can also help us to make these things consistent. We could match with `(seqable !xs ...)` and rewrite into `[!xs ...]`.

I did this work partially to understand the various parts of meander and to see what it would take to add a more significant feature like this. I have tried to test things out and ensure they work, but there could be some corner cases.

I'd definitely love to get some feedback on the idea and/or the implementation of it.

### CODE REVIEW

File: src/meander/syntax/delta.cljc [hunk 1]
```diff
@@ -514,10 +512,7 @@
                         {:pattern xs
                          :meta (meta xs)}))
         (parse
-         `(~'pred coll?
-           ;; Will cause compiler to emit a useless vector?
-           ;; check.
-           (~'app vec [~@'(_ ...) ~@(rest xs) ~@'(. _ ...)]))
+         `(~'app vec (~'seqable ~@'(_ ...) ~@(rest xs) ~@'(. _ ...)))
```
Comment (noprompt) [inline src/meander/syntax/delta.cljc:515]: We need to keep the `coll?` check around otherwise `app vec` will `throw`.

File: src/meander/syntax/delta.cljc [hunk 2]
```diff
@@ -567,6 +562,12 @@
                    :meta (meta xs)}))))
     (parse xs env)))
 
+
+(defn parse-seqable
```
Comment (noprompt) [inline src/meander/syntax/delta.cljc:566]: Needs `{:private true}` meta.

**Comment by noprompt** [comment]
I'm on board with this. I think you've done a pretty good job of making the case and it looks like you've managed to press all the right buttons in the code.

Because I want to keep my word with respect to non-breaking changes to syntax (even if it's highly improbable this would break anyone), I'd like to merge this into the `epsilon` branch if that's okay with you.

**Comment by timothypratley** [comment]
TLDR seems good

Long version:

@jimmyhmiller I'll present the case against this feature:

Given your motivating example in the tests:
```clj
(m/match [1 2 3]
         (seqable 1 2 3)
         true
         _
         false)
```

The same behavior is better achieved by:
```clj
(m/match (seq [1 2 3])
         (1 2 3)
         true
         _
         false
```

* `seq` is shorter than `seqable` (negligible value)
* `seq` is part of Clojure and preserves the match types semantics. I have control over what I pass to meander, so calling `seq` is no big chore.
* `seqable` is (to me) an overloaded concept (seqable? sequential? seq? seq). The seq, sequential, sequence abstractions are valuable in vanilla Clojure where we are limited to 10 functions that operate on 1000 data structures. They are necessary to provide a concise API for functional transformation. By contrast in Meander where we have a richer way of defining transformations; we can draw the actual shapes! This allows us to be both specific and expressive.

The "we need it" case is for nested data structures where I don't control the "subtype":

```clj
[{:a [1 2 3]} {:a (4 5 6)} {:a nil}]
```
I think `seqable` would help in this case; but I also wonder if there is a better way to approach this scenario, and where it would be possible to encounter this scenario. Just because I can't imagine it doesn't mean it wouldn't happen, maybe there are situations where this would arise. My initial feeling is that I would still prefer to handle the "sequable" part using another concept like `with` or call out to another matcher. But undeniably `sequable` would be more concise and expressive. As such it seems like a good thing to me despite my misgivings!

It would be worth documenting where/how it provides value, and pointing out that passing in a seq is an option (for example when you have a map, if you call seq on it, you can match explicitly, which would be better than using `sequable` to achieve the same thing IMO).

Sorry for the longwinded pointless rant here, just wanted to share some thinking about it and maybe prompt some outside the box thinking.

**Comment by jimmyhmiller** [comment]
No worries @timothypratley love the feedback. I definitely agree that seqable shouldn't be used as a default. And probably doesn't make sense for simple, non-nested collections. 

I definitely wouldn't call the tests cases I provided motivating examples. They were really just the simplest possible cases. The motivating example was actually implementing scan without an unnecessary check, but I do think there are other cases where seqable can come in handy. 

Like you said, seqable is primarily useful when we don't control the input and that input is nested. If I am using meander with some code that others are going to be writing, I need to make it follow general clojure practices. It isn't uncommon to have some collection that could contain mixed vectors and seqs. (This often happens when people map or filter over vectors. They could have used mapv and filterv, but often don't).

Now, of course there are other ways to handle this. It could be better to just use a `with` or an `or`. But does that express the intent that I have in this case? Personally I think no. My intent is tied up with the fact that I know the practices others follow. I want to express that my code works with anything that can be passed to me. 

If you take the last example in [my blog post](https://jimmyhmiller.github.io/meander-practical/) for example. It is using scan, which really works on seqables. Imagine instead we had written this:

```clojure 
(defn find-potential-bad-visits [data]
  (m/search data
    {:people (scan {:id ?id :name ?name})
     :addresses {?id [_ ... . {:preferred true :zip ?zip} . _ ...]}
     :visits {?id [_ ... . {:geo-location {:zip (and (not ?zip) ?bad-zip) }
                            :date ?date} . _ ...]}}
    {:name ?name
     :id ?id
     :zip ?bad-zip
     :date ?date}))
```

This would have worked until someone wrapped our visits to get rid of nil dates using a function like this:

```clojure
(defn filter-visits [visits]
  (into {} (map (fn [[key value]] [key (filter :date value)]) visits)))
```

The fact that this code would break something would come as a big surprise to the normal way of writing clojure. And in this case, my intent was really to make it work regardless of what collection type it used. 

I definitely think we need to provide guidance here. Sadly, I don't think I've used meander enough to know exactly what those best practices are yet.

**Comment by timothypratley** [comment]
@jimmyhmiller I agree that `(sequential ...)` is a necessary and nice way of choosing Clojure sequence semantics. Thank you for the extra color and examples, that helped crystallize things for me.


Tangentially I wonder if there is a nice way to explicitly choose between Clojure's map semantics and more strict map semantics. More precisely, the `associative` nil punning behavior. I might be using the wrong words here so I'll give an example:

```
(get nil :foo) -> nil
get {:foo :bar} -> bar
```

This is similar to the sequential scenario in that Clojure is fine treating `nil` as a map.
Also Meander is fine pulling `nil` out when a key is missing. But in light of using an explicit choice to use sequential vs a type for sequences, is an explicit choice for map behavior desirable? I'm having a hard time thinking through... for example if your `:visits` could be `nil` that would break things too right?

**Comment by noprompt** [comment]
@jimmyhmiller It occurs to me that we _could_ merge something like

```clj
^:seqable (,,,)
;; or
^:seqable [,,,]
```

since we do not have any special code related to these meta keys.

Considering @timothypratley's remarks as well, we could also experiment with something perhaps a bit more broad like

```clj
^:liberal {} ;; associative?
^:liberal [] ;; sequential?
^:liberal () ;; sequential?
```

Note that `nil` is neither `associative?` or `sequential?` but, if we like this direction, we could continue the discussion regarding it.

**Comment by timothypratley** [comment]
ooo nice!! I like how that keeps the seqable label outside the pattern... makes it much cleaner IMO. I also like liberal :)

**Comment by jimmyhmiller** [comment]
If we are going to introduce a new way of modifying expressions, I think we should consider some more rules about it. `^:seqable` seems weird to me. Why wouldn't we just use the type hint `^clojure.lang.Seqable` in that case? 

In general, I think we could do more with type information.

**Comment by timothypratley** [comment]
It is interesting that you bring up using an actual type.

Just to further enumerate some options here:
You could go the other direction and have a `^:strict` tag. I think the advantage of choosing your own name is that it focuses on two sensible pattern matching choices: Either by exact structure [] [] or by sequence [] ()

To extend the "use the type" line of thinking:
I think it's reasonable to describe this as choosing which interface you want to match. Do you want to match Seqable,  IPersistentCollection, IPersistentVector, ISeq, or maybe IAssociative... or maybe even a concrete type? I think that in principle going with the type hint approach allows the most expressive power here in the context of Clojure. Then there are protocols... if the tag is a protocol that the pattern satisfies, allow it. There seem to be many many tags that could be applied, but in principle, the Meander check could just be is the tag satisfied? if so, consider that to be the type check. If there is no tag, do a type check (current behavior). Is this what you have in mind?

**Comment by noprompt** [comment]
The reason I brought up using meta was to be able to use this feature on `delta`. If I'm being completely transparent, I don't like the idea of using meta but it can be useful for things like this.

---

Essentially, we want a relaxed way to match `seqable?` and, it sounds like, `associative?` things without having to pay the runtime cost or writing explicit coercion. It is possible to do something like this with `defsyntax` (pending better ergonomics).

```clj
;; We can improve defsyntax for this to be, instead, [& args].
(defsyntax seqable [pattern]
  `(~'pred seqable? (~'app seq ~pattern)))
```

The drawback with this is that we'll end up with an additional `seq?` check in the result (as I mentioned in the code). Even though the check is pretty cheap, it's wasteful and that can add up.

Since we know we _could_ support the syntax my question becomes "do we have enough information to remove the additional checks?" The check `(clojure.core/seqable? target)` will appear above the check `(clojure.core/seq? target)` in the IR so, with that knowledge, we should be able to eliminate the latter check. This reminds me of some of the [ideas](https://github.com/clojure/core.match/wiki/Crazy-Ideas) David Nolen had for `core.match`. In this case I think we could have some kind of general subsumption check i.e. `seqable?` subsumes `seq?`, `seqable?` subsumes itself, `associative?` subsumes `map?`, etc. This would help us eliminate redundant code in these cases.

**Comment by jimmyhmiller** [comment]
I do think being able to define seqable with defsyntax is the right goal. But I think what we'd want to accomplish that is a typing discipline. If we new the types of things, our check-* ops would just be noops. But I think to do that, we would want to have something less general than `pred`.

File: src/meander/syntax/delta.cljc (hunk 1 above)
Comment (jimmyhmiller) [reply src/meander/syntax/delta.cljc:515]: Good catch. I will make these changes soon. But once I land some type inference stuff, I think all of this will be unnecessary. Should land this week. Will wait to close this PR until then.

**Comment by jimmyhmiller** [comment]
Closing in favor of https://github.com/noprompt/meander/pull/50
